### PR TITLE
Make whole rail selectable

### LINKED ISSUE #415: Rails are only selectable at their arrows
### Summary
Rails are only selectable at their arrows.

### How to reproduce
1. Open any map in the editor
2. **Click on a rail (not on the arrow of the rail)**
3. Click on a rail arrow
4. **Click onto somewhere else on the rail**

### Observed behavior
2.: No rail gets selected.
4.: The selected rail gets deselected.

### Expected behavior
The rail is selected.

### Version
 - Game Version: 0.9.beta1
 - OS: Windows

### PR DESCRIPTION
**Pull request by DasCapschen**
fixes #415

### CODE REVIEW

File: src/Data/Scripts/Rail.gd [hunk 1]
```diff
@@ -58,7 +58,46 @@ func _ready() -> void:
...
+		$Arrows.queue_free()
+		$MultiMeshInstance/Area.queue_free()
+	else:
+		_generate_rail_collision()
+		_generate_direction_arrows()
+
+
+func _generate_rail_collision():
+	var child_count = $MultiMeshInstance/Area.get_child_count()
+	var mm: MultiMesh = $MultiMeshInstance.multimesh
```
Comment (HaSa1002) [inline src/Data/Scripts/Rail.gd]: It's unused. What about caching `$MultiMeshInstance/Area` instead?

File: src/Data/Scripts/Rail.gd [hunk 2]
```diff
@@ -58,7 +58,46 @@ func _ready() -> void:
...
-		$Mid.queue_free()
+		$Arrows.queue_free()
+		$MultiMeshInstance/Area.queue_free()
+	else:
+		_generate_rail_collision()
+		_generate_direction_arrows()
+
+
+func _generate_rail_collision():
+	var child_count = $MultiMeshInstance/Area.get_child_count()
```
Comment (HaSa1002) [inline src/Data/Scripts/Rail.gd]:
```suggestion
	var child_count := $MultiMeshInstance/Area.get_child_count()
```

File: src/Data/Scripts/Rail.gd [hunk 3]
```diff
@@ -58,7 +58,46 @@ func _ready() -> void:
...
+	var mm: MultiMesh = $MultiMeshInstance.multimesh
+	var col_shape_prefab = $MultiMeshInstance/Area/CS1
```
Comment (HaSa1002) [inline src/Data/Scripts/Rail.gd]:
```suggestion
	var col_shape_prefab := $MultiMeshInstance/Area/CS1 as CollisionShape
```

File: src/Data/Scripts/Rail.gd [hunk 4]
```diff
@@ -58,7 +58,46 @@ func _ready() -> void:
...
+	# left-over shapes :(
+	for i in range(1, child_count):
+		$MultiMeshInstance/Area.get_child(i).queue_free()
+
+	var step = 3
+	var box: BoxShape = col_shape_prefab.shape
+	box.extents.x = step
+	col_shape_prefab.transform.origin.x = step
+
+	for i in range(2*step, length, 2*step):
```
Comment (HaSa1002) [inline src/Data/Scripts/Rail.gd]:
```suggestion
	for i in range(2 * step, length, 2 * step):
```

File: src/Data/Scripts/Rail.gd [hunk 5]
```diff
@@ -58,7 +58,46 @@ func _ready() -> void:
...
+		var col_shape = col_shape_prefab.duplicate()
```
Comment (HaSa1002) [inline src/Data/Scripts/Rail.gd]:
```suggestion
		var col_shape := col_shape_prefab.duplicate() as CollisionShape
```

File: src/Data/Scripts/Rail.gd [hunk 6]
```diff
@@ -58,7 +58,46 @@ func _ready() -> void:
...
+		$MultiMeshInstance/Area.add_child(col_shape)
+		col_shape.owner = $MultiMeshInstance/Area
+		var offset = col_shape_prefab.transform.origin
```
Comment (HaSa1002) [inline src/Data/Scripts/Rail.gd]:
```suggestion
		var offset := col_shape_prefab.transform.origin
```

Comment (HaSa1002) [reply src/Data/Scripts/Rail.gd]: but var can also be defined out of the loop as the prefab stays the same throughout the execution

File: src/Data/Scripts/Rail.gd [hunk 7]
```diff
@@ -58,7 +58,46 @@ func _ready() -> void:
...
+		col_shape.transform = get_local_transform_at_distance(i)
+		col_shape.transform.origin += col_shape.transform.basis.xform(offset)
+
+
+func _generate_direction_arrows():
+	var positions = []
+	var count = 0
```
Comment (HaSa1002) [inline src/Data/Scripts/Rail.gd]:
```suggestion
	var positions := []
	var count := 0
```

**Review by HaSa1002 — CHANGES_REQUESTED** [review]
just gh review. Mostly style related topics. Also, please use an uppercase first letter in your commit message

File: src/Data/Scripts/Rail.gd [hunk 8]
```diff
@@ -58,7 +58,45 @@ func _ready() -> void:
...
+func _generate_rail_collision():
+	var area := $MultiMeshInstance/Area
+	var child_count := area.get_child_count()
+	var col_shape_prefab: CollisionShape = $MultiMeshInstance/Area/CS1
+	var offset := col_shape_prefab.transform.origin
+
+	for i in range(1, child_count):
+		area.get_child(i).queue_free()
+
+	var step := int(clamp(length/100, 3, 10))
```
Comment (HaSa1002) [inline src/Data/Scripts/Rail.gd]:
```suggestion
	var step := int(clamp(length / 100, 3, 10))
```

**Review by HaSa1002 — CHANGES_REQUESTED** [review]
still not a tested review

**Comment by nalquas** [comment]
Feature seems to work as intended. The regular direction arrows are a nice touch.

However, as I tried saving in the editor (on a new map), i got the following error:
`Attempt to call function 'get_child_count' in base 'null instance' on a null instance.` with the following stack trace:
Screenshot_20220920_153940

**Comment by DasCapschen** [comment]
> Feature seems to work as intended. The regular direction arrows are a nice touch.
> 
> However, as I tried saving in the editor (on a new map), i got the following error: `Attempt to call function 'get_child_count' in base 'null instance' on a null instance.`


I cannot replicate this issue. Though your stack-trace suggests to me you are using the old chunk system. That might be the issue.

I think this is ready to be merged.

**Review by nalquas — APPROVED** [review]
> > Feature seems to work as intended. The regular direction arrows are a nice touch.
> > However, as I tried saving in the editor (on a new map), i got the following error: `Attempt to call function 'get_child_count' in base 'null instance' on a null instance.`
> 
> I cannot replicate this issue. Though your stack-trace suggests to me you are using the old chunk system. That might be the issue.

Yup, can't reproduce that error anymore. You probably fixed it with the rebase.
lgtm

**Review by HaSa1002 — APPROVED** [review]
lgtm

**Pull request merged**